kratos/modeler/nurbs_volume_grid_modeler.h

```diff
@@ -0,0 +1,106 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Manuel Messmer
+//
+
+#if !defined(KRATOS_NURBS_VOLUME_GRID_MODELER_H_INCLUDED )
+#define  KRATOS_NURBS_VOLUME_GRID_MODELER_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "modeler/modeler.h"
+#include "geometries/nurbs_volume_geometry.h"
+#include "geometries/nurbs_shape_function_utilities/nurbs_volume_utilities.h"
+#include "includes/node.h"
+#include "includes/model_part.h"
+
+namespace Kratos {
+
+class KRATOS_API(KRATOS_CORE) NurbsVolumeGridModeler
+    : public Modeler
+{
+public:
+    ///@name Type Definitions
+    ///@{
+    KRATOS_CLASS_POINTER_DEFINITION( NurbsVolumeGridModeler );
+
+    typedef std::size_t IndexType;
+    typedef std::size_t SizeType;
+    typedef Node<3> NodeType;
+
+    typedef NurbsVolumeGeometry<PointerVector<NodeType>> NurbsVolumeGeometryType;
+    typedef typename NurbsVolumeGeometryType::Pointer NurbsVolumeGeometryPointerType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    NurbsVolumeGridModeler()
+        : Modeler() {}
+
+    /// Constructor.
+    NurbsVolumeGridModeler(
+        Model & rModel,
+        const Parameters ModelerParameters = Parameters())
+        : Modeler(rModel, ModelerParameters)
+        , mpModel(&rModel)
+    {
+    }
+
+    /// Destructor.
+    virtual ~NurbsVolumeGridModeler() = default;
+
+    /// Creates the Modeler Pointer
+    Modeler::Pointer Create(Model& rModel, const Parameters ModelParameters) const override
+    {
+        return Kratos::make_shared<NurbsVolumeGridModeler>(rModel, ModelParameters);
+    }
+
+    ///@}
+    ///@name Stages
+    ///@{
+
+    void SetupGeometryModel() override;
+
+    void SetupModelPart() override;
```

Comment:
I would make this modeler completely within SetupGeometryModel().

My arguments are:
- the split is theoretically logical, however, you want to be sure that no one touches this geometry in the meantime. When you split it it might happen that someone analyses the nodes and creates dependencies or something which could lead to errors.
- logically, the split is not required. If splitted it would probably involve two different modelers. One that defines the grid and one that does refinement. But I don't think that this should be the case, here.